internals/overlord/overlord.go

```diff
@@ -321,6 +328,25 @@ func (o *Overlord) ensureTimerSetup() {
 	o.pruneTicker = time.NewTicker(pruneInterval)
 }
 
+func (o *Overlord) ensureTimerStopClearChannel() {
+	o.ensureLock.Lock()
+	defer o.ensureLock.Unlock()
+	// The Go v1.x timer implementation does not guarantee that a Reset
+	// clears the channel. The documented proposal is to call stop and
+	// manually clear the channel if it already expired. This is required
+	// on loop entry because the timer is started much earlier, and
+	// managers are now allowed to call state.EnsureBefore() as soon as state
+	// is available. If the ensure timer expired before loop entry, this code
+	// ensures the channel is cleared and that only the ensure on entry is
+	// performed (and not a second ensure as a result of the channel).
+	// See: https://pkg.go.dev/time#Timer.Reset
+	o.ensureTimer.Stop()
+	select {
+	case <-o.ensureTimer.C:
+	default:
+	}
```

Comment:
Out of interest, why don't we use the pattern documented in the `time` package docs? Would this do the same thing?

```go
if !t.Stop() {
	<-t.C
}
```

Reply:
Having a line that could hard block like that (brick the code), and trusting the implementation of a standard library not to change, felt like something that would prevent me from sleeping :)